google-cloud-storage/src/main/java/com/google/cloud/storage/MultipartUploadClientImpl.java

```diff
@@ -0,0 +1,153 @@
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.storage;
+
+import com.fasterxml.jackson.dataformat.xml.XmlMapper;
+import com.google.api.client.http.HttpRequestFactory;
+import com.google.api.client.http.HttpResponse;
+import com.google.cloud.storage.Retrying.Retrier;
+import com.google.cloud.storage.multipartupload.model.CreateMultipartUploadRequest;
+import com.google.cloud.storage.multipartupload.model.CreateMultipartUploadResponse;
+import com.google.common.net.MediaType;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URI;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.text.SimpleDateFormat;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.TimeZone;
+
+/**
+ * This class is an implementation of {@link MultipartUploadClient} that uses the Google Cloud
+ * Storage XML API to perform multipart uploads.
+ */
+public final class MultipartUploadClientImpl extends MultipartUploadClient {
+
+  private final MultipartUploadHttpRequestManager httpRequestManager;
+  private final XmlMapper xmlMapper;
+  private final HttpStorageOptions options;
+  private final Retrier retrier;
+  private final URI uri;
+
+  public MultipartUploadClientImpl(
+      URI uri, HttpRequestFactory requestFactory, Retrier retrier, HttpStorageOptions options) {
+    this.httpRequestManager = new MultipartUploadHttpRequestManager(requestFactory);
+    this.xmlMapper = new XmlMapper();
+    this.options = options;
+    this.retrier = retrier;
+    this.uri = uri;
+  }
+
+  private Map<String, String> getGenericExtensionHeader() {
+    Map<String, String> extensionHeaders = new HashMap<>();
+    if (options.getClientLibToken() != null) {
+      extensionHeaders.put("x-goog-api-client", options.getClientLibToken());
+    }
+    if (options.getProjectId() != null) {
+      extensionHeaders.put("x-goog-user-project", options.getProjectId());
+    }
+    extensionHeaders.put("Date", getRfc1123Date());
+    return extensionHeaders;
+  }
+
+  public CreateMultipartUploadResponse createMultipartUpload(CreateMultipartUploadRequest request)
+      throws IOException {
+    String encodedBucket = encode(request.bucket());
+    String encodedKey = encode(request.key());
+    String resourcePath = "/" + encodedBucket + "/" + encodedKey;
+    String createUri = uri.toString() + resourcePath + "?uploads";
```

Comment:
I think we've got a bit of split brain happening between this class and MultipartUploadHttpRequestManager. This class is doing url encoding and construction, header processing and response parsing. But, Manager is doing some http request initialization, request execution and response returning.

I'd expect the logic of building the actual http request to happen over in manager, while this class would focus on any retry wrapping and additional info that would need to be provided in order for the manager to actually process an individual request. (for example, manager doesn't do anything with the provided CreateMultipartUploadRequest provided to it).

What is the intended delineation of responsibility between these two classes?

Reply:
The idea was for MultipartUploadClientImpl to be responsible for extracting all arguments from CreateMultipartUpload request and passing it to manager to perform actual httpRequest stuff.

MultipartUploadHttpRequestManager should be responsible for doing the actual requestFactory and creating httpRequest stuff.

I am fine with moving extraction part to manager as well, as we are anyways sending the request for some of the direct headers.